### PR TITLE
Improve `GetNavigationSource` performance

### LINKED ISSUE #1118: Perf: Optimize `GetNavigationSource(ODataPath)`
The `GetNavigationSource(ODataPath)` is relatively expensive. Part of the cost is due to the fact that we rely on a custom path segment handler to determine the path.

Once we ship an ODL version that support indexing ODataPath, we can rewrite this method using a simple loop that reads the segments from the end until it find the appropriate one to determine the navigation source.

Related:
- https://github.com/OData/odata.net/issues/2814
- https://github.com/OData/odata.net/issues/2804
- https://github.com/OData/odata.net/pull/2811

### PR DESCRIPTION
**Pull request by habbes**
Fix #1118 

Replaces the use of `ODataPathNavigationSourceHandler` in `ODataPath.GetNavigationSource` extension method with a simpler reverse loop of the path.

With `ODataPathNavigationSourceHandler`
- a new instance of the handler is allocated
- all the segments of the path are visited
- a `List<string>` is created and populated with the string representation of each path segment. This list is used to generate a string representation of the path via a `Path` property. But this property is not needed to determine the navigation source (it's also not used by existing code).

The new implementation scans the path segments from the end and stops as soon as it finds the navigation source because only the last navigation source in the path matters.

Also, the `ODataPath.GetNavigationSource()` method didn't have good test coverage, so I added tests to verify all the scenarios the original implementation handled.

The `ODataPathNavigationSourceHandler` is public so I kept it to avoid breaking changes.

**Before**

image

**After**

image

### CODE REVIEW

File: src/Microsoft.AspNetCore.OData/Routing/ODataPathExtensions.cs [hunk 1]
```diff
@@ -77,9 +77,59 @@ public static IEdmNavigationSource GetNavigationSource(this ODataPath path)
...
+                {
+                    return typeSegment.NavigationSource;
+                }
+
+                if (segment is PropertySegment)
+                {
+                    continue;
+                }
+
+                return null;
```
Comment (wachugamaina) [inline src/Microsoft.AspNetCore.OData/Routing/ODataPathExtensions.cs:128]: So if we are unable to match one segment we abandon everything?

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs [hunk 2]
```diff
@@ -63,6 +58,88 @@ public void GetNavigationSource_ThrowsArgumentNull_Path()
...
+        [InlineData("VipCustomer/Name", "VipCustomer", EdmNavigationSourceKind.Singleton)]
+        [InlineData("VipCustomer/Orders", "Orders", EdmNavigationSourceKind.EntitySet)]
+        [InlineData("VipCustomer/NS.IsUpgraded", null, EdmNavigationSourceKind.None)]
+        [InlineData("Customers(1)/NS.SpecialCustomer", "Customers", EdmNavigationSourceKind.EntitySet)]
+        [InlineData("VipCustomer/NS.SpecialCustomer", "VipCustomer", EdmNavigationSourceKind.Singleton)]
+        [InlineData("Customers/$count", null, EdmNavigationSourceKind.None)]
+        [InlineData("Customers(1)/Account/Amount/$value", null, EdmNavigationSourceKind.None)]
+        [InlineData("$batch", null, EdmNavigationSourceKind.None)]
+        [InlineData("$metadata", null, EdmNavigationSourceKind.None)]
+        public void GetNavigationSource_ReturnsCorrectNavigationSource(string path, string expectedNavigationSource, EdmNavigationSourceKind navigationSourceKind)
```
Comment (wachugamaina) [inline test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:94]: I dont belive that all the handler cases are covered in these tests, for example I cannot see 
operationImportSegment
operationSegment
TypeSegment 
DynamicPathSegment
PathTemplateSegment 
ValueSegment 
CountSegment 
BatchSegment 
MetadataSegment 
ODataPathSegment

File: src/Microsoft.AspNetCore.OData/Routing/ODataPathExtensions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (habbes) [reply src/Microsoft.AspNetCore.OData/Routing/ODataPathExtensions.cs:128]: Yes, they should return null. That's the existing behaviour. Also handled in tests.

File: test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs (hunk 2 above)
Comment (habbes) [reply test/Microsoft.AspNetCore.OData.Tests/ODataPathExtensionsTests.cs:94]: All those scenarios are covered:

- `OperationImportSegment`: lines 78 and 79
- `OperationSegment`: lines 8-83, 87
- `TypeSegment`: 88. 89
- `DynamicPathSegment`: 77
- `ValueSegment`: 91
- `CountSegment`: 90
- `MetadataSegment`:  93
- `BatchSegment`: 92

`PathTemplateSegment` is covered separately in the test `GetNavigationSource_WhenPathTemplateSegment_ReturnsNull` on line 116

`ODataPathSegment` is the abstract base class. I tested this by creating a new `UnknownTestODataPathSegment` and verifying that it returns null in `GetNavigationSource_WhenUnkownPathSegment_ReturnsNull` on line 130.

File: src/Microsoft.AspNetCore.OData/Routing/ODataPathExtensions.cs [hunk 3]
```diff
@@ -77,9 +77,59 @@ public static IEdmNavigationSource GetNavigationSource(this ODataPath path)
                 throw Error.ArgumentNull(nameof(path));
             }
 
-            ODataPathNavigationSourceHandler handler = new ODataPathNavigationSourceHandler();
```
[the remark this answers is not in the document]
Comment (habbes) [reply src/Microsoft.AspNetCore.OData/Routing/ODataPathExtensions.cs:80]: Alright, let me create a separate PR to add the tests.

[the remark this answers is not in the document]
Comment (habbes) [reply src/Microsoft.AspNetCore.OData/Routing/ODataPathExtensions.cs:80]: I've created the tests PR: https://github.com/OData/AspNetCoreOData/pull/1159

**Comment by habbes** [comment]
Closing this PR since I'm splitting the test and refactoring in separate PRs. I will re-create the PR once the tests have been merged: https://github.com/OData/AspNetCoreOData/pull/1159

**Comment by habbes** [comment]
Replaced with: https://github.com/OData/AspNetCoreOData/pull/1161

**Pull request closed without merging**